posts/2024-05-01-enabling-rust-lld-on-linux.md

```diff
@@ -0,0 +1,93 @@
+with a version of `lld` for this purpose. When CI builds LLVM to use in the compiler, it also builds
+the linker and packages it. It's referred to as `rust-lld` to avoid colliding with any `lld` already
+installed on the user's machine.
+
+Since improvements to linking times are substantial, it would be a good default to use in the most
+popular targets. This has been discussed for a long time, for example in issues
+[#39915](https://github.com/rust-lang/rust/issues/39915) and
+[#71515](https://github.com/rust-lang/rust/issues/71515), and rustc already offers nightly flags to
+use `rust-lld`.
+
+By now, we believe we've done all the internal testing that we could, on CI, crater, our
+benchmarking infrastructure, and would like to expand testing and gather real-world feedback and
+use-cases. Therefore, we will enable `rust-lld` to be the linker used by default on
+`x86_64-unknown-linux-gnu` for nightly builds.
+
+#### Benefits
+
+While this also enables the compiler to use more linker features in the future, the most immediate
+benefit is much improved linking times.
+
+Here are more details from the ripgrep example mentioned above: linking is reduced 7x, resulting in
+a 40% reduction in end-to-end compilation times.
+
+![Before/after comparison of a `ripgrep` debug build](../../../../images/2024-05-01-enabling-rust-lld-on-linux/ripgrep-comparison.png)
+
+Most binaries should see some improvements here, but it's especially significant with e.g. bigger
+binaries, or when involving debuginfo. These usually see bottlenecks in the linker.
+
+Here's [a
+link](https://perf.rust-lang.org/compare.html?start=b3e117044c7f707293edc040edb93e7ec5f7040a&end=baed03c51a68376c1789cc373581eea0daf89967&stat=instructions%3Au&tab=compile)
+to the complete results from our benchmarks.
+
+If testing goes well, we can then stabilize using this faster linker by default for
+`x86_64-unknown-linux-gnu` users, before maybe looking at other targets.
+
+#### Possible drawbacks
+
+From our prior testing, we don't really expect issues to happen in practice. It is a drop-in
+replacement for the vast majority of cases, but `lld` is not _bug-for-bug_ compatible with GNU ld.
+
+In any case, using `rust-lld` can be disabled if any problem occurs: use the `-Z
+linker-features=-lld` flag to revert to using the system's default linker.
+
+Some crates somehow relying on these differences could need additional link args. For example, we
+saw <20 crates in the crater run failing to link because of a different default about [encapsulation
+symbols](https://lld.llvm.org/ELF/start-stop-gc): these could require
+`-Clink-arg=-Wl,-z,nostart-stop-gc` to match the legacy GNU ld behavior.
+
+Some of the big gains in performance come from parallelism, which could be undesirable in
+resource-constrained environments.
+
+#### Summary
+
+rustc will use `rust-lld` on `x86_64-unknown-linux-gnu` nightlies, for much improved linking times.
```

Comment:
Are there any plans to make a similar for windows/macos? It might be good to link to tracking issues for these if they exist.

Reply:
There is a tracking issue for windows in https://github.com/rust-lang/rust/issues/71520 that is also linked in the tracking issues linked in the post, but I personally don't know of concrete plans to make this happen in the near future:
- I presume lld is a lot more tested on linux. There are also known miscompilations when using rustc/ThinLTO/lld on windows, that currently prevent us from building rustc with ThinLTO there, for example.
- it's much harder for us to test such a change, we can't currently do crater runs for windows.
- last I checked, the gap between link.exe and lld on windows was much smaller than between GNU ld and lld on linux.

For mac, it also seems less useful since Apple released their own fast linker to replace `ld64`, and we don't have a tracking issue for switching the linker there.

The post discusses "doing this on linux before maybe looking at other targets" and that seems like an accurate summary, given the above.